### ISSUE TITLE
Filterning terrains without overlay doesn't work as documented

### ISSUE THREAD

**Issue by ProditorMagnus** [wesnoth/wesnoth#7734]
### Game and System Information

- **Version:** 1.16.0

### Description of the bug

`terrain=*` matches terrains with overlay, while it is documented that it should not.

### Steps to reproduce the behavior

https://forums.wesnoth.org/viewtopic.php?p=682310#p682310

### Expected behavior

Decide correct behaviour, and either update documentation or game.

**Comment by ProditorMagnus** [comment]
I start to favor * matching all.

**Comment by CelticMinstrel** [comment]
I think if `*` matches all and `*^*` matches only terrains with overlay, you could still achieve "terrains without overlay" via `!,*^*`.

**Comment by ProditorMagnus** [comment]
Their issue was that `*` and `*^*` are equal filter.

**Comment by CelticMinstrel** [comment]
> It would be bad idea to break compatibility by making this overlay only.

I disagree. If necessary we could add extra keys (maybe `strict=yes`) to preserve backwards compatibility, but I think making `*^*` only match terrains with overlay is correct. The pattern implies that it's looking for something with a `^` in it – matching things without one is nonsensical.

**Comment by ProditorMagnus** [comment]
| | Current| Currently documented| CelticMinstrel suggestion|
|--------|--------|--------|--------|
| `*`| any terrain| terrain without overlay| any terrain|
| `*^*`| any terrain| any terrain| terrain with overlay|

It would swap the expected meanings. Having key to ask for opposite behaviour doesnt look good.

**Comment by CelticMinstrel** [comment]
It would also make a thousand times more sense. I think the deprecation process of swapping the meaning from what's documented would be worth it. Plus it doesn't even work as documented in the first place, so the deprecation process wouldn't actually be swapping the meaning – it would make it work as it _actually_ does in 1.16 if you don't use that key, rather than how it's documented to work.

**Comment by ProditorMagnus** [comment]
If deprecation messages can detect changed situations I am ok with the suggested change. So for example `*` and `Gg^*` should have deprecation message, and `*^V*` should not have?

I still dont think key would help at all. New style for terrain without overlay would be `!,*^*,!`
If you know you need change which could be achieved with key, you can just update filter.

**Comment by CelticMinstrel** [comment]
The purpose of the key is to provide a transitional period. It would go like this:

* **1.18**: Default behaviour is `strict=no`, same as 1.16. When updating the filter you are expected to add `strict=yes`. If the `strict` key is missing there will be a deprecation warning that the behaviour is going to change in 1.20, so if you don't want to update it right away you should add `strict=no`.
* **1.20**: Default behaviour is now `strict=yes`, but it still uses old 1.16 behaviour if it sees `strict=no`, with a deprecation warning that this behaviour is going away and what to do about it.
* **1.22**: The `strict` key is no longer supported; always use the new behaviour.

**Comment by ProditorMagnus** [comment]
I understand the concept now. I dont think there is enough reason to have key, making change in behavior with 1.19.0 would give enough opportunity to update filters if needed.

**Comment by doofus-01** [comment]
There have been other issues with terrain filtering, such as `Ira` matching `Ia*`.  That's from memory, I'll try to come up with a confirmed list, but just wanted to mention it now in case this change takes off soon.

I'm assuming terrain graphics and [filter_location] (or whatever was the origin of the OP) use the same rules.  If they don't, they should.

**Comment by CelticMinstrel** [comment]
I think this would affect terrain graphics and anything that uses a standard location filter. I'm not sure if there's anywhere else? We should probably make sure.

**Comment by doofus-01** [comment]
Testing now (in the editor, modifying one of the EDITOR_DEPRECATED lines in `core/terrain-graphics.cfg`), I see I wasn't quite remembering correctly, but it's probably still worth mentioning:
1.  `I*a` catches `Isr` (i.e. `I*a` is `I*`), does not catch `Gg`
2.  `I*a,!,*^Emf` catches everything without `Emf` overlay (`Gg` for example), catches `I*^Emf`, does not catch `Gg^Emf`
3. `!,*^Emf,!,I*a` catches only `I*` without the `^Emf` overlay

Item 1 seems like a bug or at least an unintuitive limitation.  I don't think items 2 and 3 are bugs, and changing the behavior will require reworking the terrain graphics rules, but more examples in the wiki documentation would be nice.

**Comment by CelticMinstrel** [comment]
Perhaps for the `!` situation it would make sense to add a new syntax for doing that in a way that makes more sense. We could use `~` maybe? Perhaps something like `I*a~*^Emf` matches all `I*a` that don't have the `^Emf` overlay? That probably merits a separate enhancement request though.

From what I'm seeing, I think the issue is that terrain string globbing is not using glob logic but rather some weird special logic. I think just using glob directly would solve both the stated issues – the case for `*` / `*^*` and the case for `I*a`.

**Comment by ProditorMagnus** [comment]
I dont support ~ or even !, those use cases should be handled with [and][not][or] tags.

**Comment by CelticMinstrel** [comment]
But `[and][or][not]` tags aren't supported in terrain graphics, so something is needed at least for that use-case. And using a full location filter in terrain graphics feels like it might be putting the cart before the horse.

**Comment by doofus-01** [comment]
`[and][or][not]` tags in terrain graphics would be a big project and make the terrain graphics rules even less readable than they already are.

I think the current ! notation is OK, unless there is some case that is inaccessible.  I have not run into that yet, but it would be interesting to find out if there was a case.  Until then, I'm not convinced a new syntax is worth the effort.  Current one just needs to be documented better.

The current * notation quirks, on the other hand, don't seem useful at all.  Are the `I*a` and `*^*` issues really the same thing?  If the root cause is that the engine is just tossing out everything after the first * it comes across, that would be consistent with our current results, and seems like fixing that wouldn't involve new syntax or deprecation.

**Comment by CelticMinstrel** [comment]
> I think the current ! notation is OK, unless there is some case that is inaccessible. I have not run into that yet, but it would be interesting to find out if there was a case. Until then, I'm not convinced a new syntax is worth the effort. Current one just needs to be documented better.

I think the `!` notation is a total hack and should be replaced with something better if we can think of it (while still supporting `!` for compatibility). Including a `!` as a list element to somehow control the meaning of subsequent elements is a terrible idea, and as you've said yourself, the way it works is confusing. Even your attempt at explaining how it works doesn't make any sense:

> "!" behavior can be confusing if you think of it like a toggle (an idea you could get from the wiki currently). "!,Xx,!,Yy" is not "!Xx,Yy", it is "!(Xx,!Yy)"

What is that even supposed to mean? I guess the main problem is that you seem to be attempting to explain the syntax in terms of itself, but I'm still confused after reading it.

**Comment by ProditorMagnus** [comment]
Created unit test to confirm X*y behaviour. I do not rely on such filters, or any nontrivial "!," so not going to focus on getting such test to core standards. 

```
# 

#####
# API(s) being tested: 
##
# Expected end state:
# 
#####
{GENERIC_UNIT_TEST "terrain_filter_wildcard" (
    [event]
        name = start
        [lua]
            code = <<
local run_test_case = function (filter_string, matching_terrain, invalid_terrain)
	if #matching_terrain + #invalid_terrain > 16 then
		unit_test.fail()
	end
	
	local expected_match_index = {}
	local x = 0
	for i, t in ipairs(matching_terrain) do
		x = x + 1
		wesnoth.current.map[{x, 1}] = t
		--wesnoth.wml_actions.terrain{terrain=t, x=x, y=1}
		expected_match_index[tostring(x)] = true
	end
	for i, t in ipairs(invalid_terrain) do
		x = x + 1
		wesnoth.current.map[{x, 1}] = t
		--wesnoth.wml_actions.terrain{terrain=t, x=x, y=1}
	end
	
	-- all find output are part of matching_terrain
	for i, loc in ipairs(wesnoth.map.find{terrain=filter_string, x="1-"..tostring(x), y="1"}) do
		
		unit_test.assert_equal(expected_match_index[tostring(loc.x)], true, filter_string.." should not match "..loc.terrain)
		expected_match_index[tostring(loc.x)] = false
	end
	
	-- none from matching_terrain is missing from find output
	for k, v in pairs(expected_match_index) do
		unit_test.assert_equal(v, false, filter_string.." should match "..wesnoth.map.find{x=k, y="1"}[1].terrain)
	end
end

run_test_case("*", {"Gg", "Gg^Wm"}, {})
run_test_case("*^*", {"Gg", "Gg^Wm"}, {})
run_test_case("W*g", {"Wog", "Wwg"}, {"Wot"})
run_test_case("Wo*", {"Wog", "Wot"}, {"Wwg"})

run_test_case("!,Wo*", {"Wwg"}, {"Wog", "Wot"})
run_test_case("!,Wo*,Wog", {"Wwg"}, {"Wog", "Wot"})
run_test_case("Wog,!,Wo*", {"Wog", "Wwg"}, {"Wot"})


unit_test.succeed()
            >>
        [/lua]
    [/event]
)}
```
Assertion failed (expected nil == true): W*g should not match Wot

Designed to check terrain filter string against provided examples it should or should not match.

**Comment by doofus-01** [comment]
>  Including a ! as a list element to somehow control the meaning of subsequent elements is a terrible idea, and as you've said yourself, the way it works is confusing. Even your attempt at explaining how it works doesn't make any sense:

This page had the most detailed (though confusing to me) explanation of the ! notation: https://wiki.wesnoth.org/StandardLocationFilter#Filtering_Terrains

Checking the history, to see how long it had been that way, it looks like it was written five years ago by a Celtic Minstrel.

Here's another attempt:
 - `,` means OR
 - `!` means NOT
 - `!,Xx,!,Yy` means NOT( Xx OR (NOT(Yy)))
 - An alternative (and wrong) interpretation of `!,Xx,!,Yy` could have been (NOT(Xx)) OR (Yy), as if each `!` was flipping a negation switch on/off, or as if every odd/even `!` was an opening/closing tag ([not]...[/not]).  But `!` just negates the result of everything to the right, which can include other `!`, until the end of the list.

The "!" does what it needs to, as far as I can tell, and it wouldn't be confusing if it were explained accurately somewhere.  I'm not in favor of changing it for aesthetics.  If there was some new capability that could simplify the existing rules and make better new ones, that would be a different story.

The "*" on the other hand, really does have room for improvement.  Changing the behavior will surely involve some broken terrain graphics, but I don't think there is anything that really requires the current behavior.

**Comment by CelticMinstrel** [comment]
I think that explanation makes sense. I still don't like the syntax, but I wasn't suggesting to get rid of it anyway, just add something new that most of the time does the job better.

**Comment by doofus-01** [comment]
Fair enough.  Perhaps a new negation method would make sense in conjunction with expanded hex data (such as elevation, or other things like weather markers) that might need more sophisticated inline filtering.   For the existing "!", I'll update the wiki this weekend.  Thanks.

(BTW, I didn't mean the earlier wiki history thing as a slap, I was partially checking it to make sure I wasn't the author - five years is a long time.)

**Comment by ProditorMagnus** [comment]
Update for 1.19.10 since I accidentally noticed `**` has another meaning.

| | Current| Currently documented| CelticMinstrel suggestion|
|--------|--------|--------|--------|
| `*`| any terrain| terrain without overlay| any terrain|
| `*^*`| any terrain| any terrain| terrain with overlay|
| `**`| terrain with overlay| not documented| ?|

It would swap the expected meanings. Having key to ask for opposite behaviour doesnt look good.

**Comment by soliton-** [comment]
Short circuit of `*`:
https://github.com/wesnoth/wesnoth/blob/9cc5911bc81e3ab8baecb6c1073b2cadc4c8c81b/src/terrain/translation.cpp#L527-L528

Without short circuit base and overlay are checked:
https://github.com/wesnoth/wesnoth/blob/9cc5911bc81e3ab8baecb6c1073b2cadc4c8c81b/src/terrain/translation.cpp#L543-L546

Thus `*` and `**` behave differently. This is confusing IMO. Both should behave the same. And it makes sense to me that similar to pathname expansion (globbing) the wildcard is only applied to one path component or in the terrain string case one layer.

To lower the chance of breaking existing content we could make it so that only `*^` (currently parsed the same as `*`) means no overlay. At least I expect that to not be meaningfully used right now.

**Comment by soliton-** [comment]
There is a single `*^$var` use in 1.18/1.19 addons and mainline terrain graphics uses `*^{MACRO_ARG}`. I expect neither is supposed to expand to `*^` so it looks like `*^` is indeed not used right now.

**Comment by soliton-** [comment]
@doofus-01 as probably our heaviest user of terrain graphics you may know that `[terrain_graphics] [tile] type=Gg` is interpreted as `[terrain_graphics] [tile] type=Gg^*`. This means if you ever want to just match a terrain with no overlay you simply cannot.
1. Has this ever bothered you?
2. Would it be a problem if you actually needed to write `Gg^*` if you do want to match with any overlay? (Right now it appears mainline terrain graphics WML very often leaves the `^*` part out.)

**Comment by doofus-01** [comment]
@soliton- 
1.  It hasn't really come up, mostly because overlay can mean almost anything, from a little flower or sunbeam to a house or forest.  Usually if there is some overlay filtering needed, it relates to specific overlays.  That makes it important to follow the terrain code conventions, but the sloppy `^*` interpretation hasn't caused trouble, as far as I know.
2. I see the logic and can't immediately see any issue with it, assuming the existing usages are all updated appropriately.  I think it will end up meaning `Gg` becomes `Gg,Gg^*` most of the time, which I guess could be annoying if we try to avoid baking that into the macros.
 
In the end, I see what you mean and why you want to fix it, but I'm not sure this will help much and could create a minor PITA going forward.  Can we hold off changing it until someone has a clear example of where the existing quirkiness is blocking something?  There could very well be something, but I can't think of it right now.

**Comment by soliton-** [comment]
(`Gg` would become `Gg^*`) It’s a bit unfortunate that the "Ugly hack" (actual code comment) from years ago led to this but we can certainly also keep the exceptional behavior in terrain graphics for now and if we can find no sensible transition path also document it.

I did suspect that it's rare to want to match some base terrain without any overlay in terrain graphics. Actually my bad on saying that you cannot match a (specific) terrain without overlay though. While `*^` acts the same as `*^*` that is a special case. In general `Gg^` does mean grassland with no overlay even in terrain graphics. This is not used in current mainline terrain graphics but was used in earlier versions and is used in UMC.
